Decouple periodic events from controller names

kdmnk opened the pull request:
Now, the controllers using a periodicEvent can set which periodicEvent do they connected to.


The Controllers that use a periodicEvents can define which periodicEvent they want to edit and use in their constructor. The possible options can be found in the PeriodicEvents.
When we handle the start/end dates, we fire events and use event handlers.

Note: I had to move some functions out from the controllers so that the event handlers can take care of the functionalities.

kdmnk committed:
Decouple periodic events from controller names

kdmnk committed:
Use Laravel events instead of retrieving classes

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 42074e1 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/555

kdmnk committed:
rename event_model to event_name

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in 65ca69f according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/555

kdmnk commented on app/Models/PeriodicEvent.php:
```diff
@@ -112,13 +127,19 @@
+                break;
+            default:
+                Log::debug("No event handler defined for ". $this->event_model . "'s start time");
+        }
```
Can you give an example solution?

kdmnk committed:
cleanup

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in e1211b4 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/555

kdmnk committed:
fix test

kdmnk committed:
Merge branch 'development' into decouple-periodic-events

kdmnk committed:
Merge branch 'development' into decouple-periodic-events

kdmnk committed:
fix phpstan issues

kdmnk committed:
Merge remote-tracking branch 'origin/development' into decouple-periodic-events

kdmnk committed:
fix admmissioncontroller

kdmnk committed:
fix tests

viktorcsimma commented on app/Listeners/SemesterEvaluationPeriodStartListener.php:
```diff
@@ -0,0 +1,28 @@
+        }
+        if (User::president()) {
+            Mail::to(User::president())->queue(new EvaluationFormAvailableDetails(User::president()->name, $event->periodicEvent->deadline()));
+        }
```
Should we add the director for this, too? He appears in the end listener.

viktorcsimma commented on app/Models/PeriodicEvent.php:
```diff
@@ -112,13 +121,19 @@ public function isExtended(): bool
     {
-        //Get the corresponding controller and call its start method
-        app($this->event_model)->handlePeriodicEventStart();
+        switch ($this->event_name) {
+            case self::SEMESTER_EVALUATION_PERIOD:
+                event(new SemesterEvaluationPeriodStart($this));
+                break;
```
This seems not to be open-closed; can we circumvent this somehow? Maybe through singletons derived from this class.

kdmnk replied:
Yeah I remember that I was trying a lot of stuff here. We can make this more fancier but I don't see the point of making this more complex, and with registers or whatever else we still need to do add some code manually (it was calculated automatically before, but now the point of this is to decouple the names from controllers, so the mapping has to be manual somewhere), just in a more abstracted, harder to find place. 

But I'm curious to see if you find a better way to structure this, or move it to somewhere that makes more sense to you.

viktorcsimma commented on app/Models/PeriodicEvent.php:
```diff
@@ -141,10 +161,13 @@ public function handleEnd(): void
-
-        //Get the corresponding controller and call its start method
-        app($this->event_model)->handlePeriodicEventReminder($days_left);
+        switch ($this->event_name) {
+            case self::SEMESTER_EVALUATION_PERIOD:
+                event(new SemesterEvaluationPeriodReminder($this));
+                break;
+            default:
```
Same here.

viktorcsimma commented on app/Utils/PeriodicEventController.php:
```diff
@@ -56,20 +65,20 @@ final public function periodicEventForSemester(?Semester $semester): ?PeriodicEv
             ->first();
     }
 
     /**
-     * Create or update the current PeriodicEvent connected to the model.
-     * Make sure the $data is properly validated:
-     * @param Semester $semester
-     * @param Carbon $start_date
-     * @param Carbon $end_date
-     * @param Carbon|null $extended_end_date
-     * @return PeriodicEvent
-     */
+    * Create or update the current PeriodicEvent connected to the model.
+    * Make sure the $data is properly validated:
+    * @param Semester $semester
+    * @param Carbon $start_date
+    * @param Carbon $end_date
+    * @param Carbon|null $extended_end_date
+    * @return PeriodicEvent
+    */
```
Nitpick, but I think the asterisks were indented right.

viktorcsimma commented on app/Http/Controllers/Secretariat/SemesterEvaluationController.php:
```diff
@@ -141,20 +74,20 @@ public function show()
             'faculties' => Faculty::all(),
             'workshops' => Workshop::all(),
-            'evaluation' => user()->semesterEvaluations()->where('semester_id', Semester::current()->id)->first(),
+            'evaluation' => user()->semesterEvaluations()->where('semester_id', self::semester()->id)->first(),
```
@kdmnk this is indeed true; what about this?
```php
where('semester_id', $this->periodicEvent()?->semester()->id ?? Semester::current())
```

kdmnk replied:
self::semester() is defined in parent class. I made it nullable though. Semester::current() might introduce confusion again.

viktorcsimma commented on app/Http/Controllers/Secretariat/SemesterEvaluationController.php:
```diff
@@ -141,20 +74,20 @@ public function show()
+            'evaluation' => user()->semesterEvaluations()->where('semester_id', self::semester()->id)->first(),
             'general_assemblies' => GeneralAssembly::all()->sortByDesc('closed_at')->take(2),
-            'community_services' => user()->communityServiceRequests()->where('semester_id', Semester::current()->id)->get(),
+            'community_services' => user()->communityServiceRequests()->where('semester_id', self::semester()->id)->get(),
```
Same here.

viktorcsimma commented on app/Http/Controllers/Secretariat/SemesterEvaluationController.php:
```diff
@@ -141,20 +74,20 @@ public function show()
             'position_roles' => user()->roles()->whereIn('name', Role::STUDENT_POSTION_ROLES)->get(),
             'periodicEvent' => $this->periodicEvent(),
-            'users_havent_filled_out' => user()->can('manage', SemesterEvaluation::class) ? $this->usersHaventFilledOutTheForm() : null,
+            'users_havent_filled_out' => user()->can('manage', SemesterEvaluation::class) ? User::doesntHaveStatusFor(self::semester()->succ())->get() : null,
```
And here:
```php
($this->periodicEvent()?->semester()->id ?? Semester::current())->succ()
```

viktorcsimma reviewed (changes requested):
Sorry for being so late with reviewing this. Seems to be a nice change; more understandable, clear and useful:)

Besides the things I have written (for me, the semester evaluation page crashed because of `self::semester()`, there was one more thing, seen when I switched to another periodic event:

![grafik](https://github.com/user-attachments/assets/f40717d9-0af0-4cea-8e41-bf6ecfbee280)

It's only 21:48, but it is still open. Maybe my fix was incorrect.

Also, can you check out CodeRabbit comments as well?

kdmnk committed:
requested changes

kdmnk committed:
Merge remote-tracking branch 'origin/development' into decouple-periodic-events

kdmnk committed:
deactivating students through cli

deepsource-autofix[bot] committed:
style: format code with PHP CS Fixer

This commit fixes the style issues introduced in b8292c8 according to the output
from PHP CS Fixer.

Details: https://github.com/EotvosCollegium/mars/pull/555

kdmnk committed:
fix test

kdmnk commented:
Sorry for this that it took so long. Now it also supports the deactivation through the command line that was introduced in https://github.com/EotvosCollegium/mars/pull/681. 

However, I have a docker bug now so I can not test it at the moment manually.

kdmnk committed:
fix test

kdmnk committed:
fix test
